File: frontend/src/services/mood.ts
import { api } from '@/services/api'

export const EMOTIONS = [
  'gioia', 'fiducia', 'paura', 'sorpresa',
  'tristezza', 'disgusto', 'rabbia', 'anticipazione',
] as const

export type Emotion = (typeof EMOTIONS)[number]

export interface TurnRead {
  id: number
  created_at: string
  emotions: Record<string, number>
  dominant_emotions: [string, number][]
  valence: number
  arousal: number
  dominance: number
  confidence: number
  label: string
  no_lexicon_match: boolean
}

export interface FeedbackRead {
  emotions: Record<string, number>
  dominant_emotions: [string, number][]
  valence: number
  arousal: number
  dominance: number
  label: string
  feedback_count: number
  consolidated: boolean
}

export interface TurnCreate {
  text: string
  keydown_times: number[]
  backspace_count: number
  response_latency_s: number
  is_followup: boolean
  followup_depth: number
  hour_of_day?: number | null
}

export const moodApi = {
  ingestTurn: (userId: number, body: TurnCreate) =>
    api.post<TurnRead>(`/users/${userId}/turns`, body),
  listTurns: (userId: number) => api.get<TurnRead[]>(`/users/${userId}/turns`),
  sendFeedback: (userId: number, turnId: number, corretto: boolean, emozione_corretta?: Emotion) =>
    api.post<FeedbackRead>(`/users/${userId}/turns/${turnId}/feedback`, {
      corretto,
      emozione_corretta: corretto ? null : emozione_corretta,
    }),
  diagnostics: (userId: number) => api.get<Record<string, unknown>>(`/users/${userId}/model`),
}

File: backend/app/schemas/mood.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class Emotion(str, Enum):
    """Le 8 emozioni di base del modello (Plutchik)."""

    gioia = "gioia"
    fiducia = "fiducia"
    paura = "paura"
    sorpresa = "sorpresa"
    tristezza = "tristezza"
    disgusto = "disgusto"
    rabbia = "rabbia"
    anticipazione = "anticipazione"


class TurnCreate(BaseModel):
    """Input di un turno: testo + segnali comportamentali."""

    text: str = Field(min_length=1, max_length=2000)
    keydown_times: list[float] = Field(
        default_factory=list,
        description="Timestamp assoluti (s) dei tasti (<2 => nessun segnale).",
    )
    backspace_count: int = Field(default=0, ge=0)
    response_latency_s: float = Field(default=5.0, ge=0)
    is_followup: bool = False
    followup_depth: int = Field(default=0, ge=0)
    hour_of_day: Optional[int] = Field(default=None, ge=0, le=23)


class TurnRead(BaseModel):
    """Stima emotiva prodotta per un turno."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    emotions: dict[str, float]
    dominant_emotions: list[tuple[str, float]]
    valence: float
    arousal: float
    dominance: float
    confidence: float
    label: str
    no_lexicon_match: bool


class FeedbackCreate(BaseModel):
    """Feedback umano sull'ultima stima."""

    corretto: bool
    emozione_corretta: Optional[Emotion] = None

    @model_validator(mode="after")
    def _valida_emozione(self) -> "FeedbackCreate":
        if not self.corretto and self.emozione_corretta is None:
            raise ValueError("emozione_corretta e' obbligatoria quando corretto=False")
        if self.corretto and self.emozione_corretta is not None:
            raise ValueError("emozione_corretta non va indicata quando corretto=True")
        return self


class FeedbackRead(BaseModel):
    """Stato emotivo corrente dopo l'applicazione del feedback."""

    emotions: dict[str, float]
    dominant_emotions: list[tuple[str, float]]
    valence: float
    arousal: float
    dominance: float
    label: str
    feedback_count: int
    consolidated: bool


class CalibrationRequest(BaseModel):
    """Calibrazione con testo proprio (richiede consenso esplicito)."""

    testo: str = Field(min_length=1)
    autore_e_soggetto_coincidono: bool

    @model_validator(mode="after")
    def _valida_consenso(self) -> "CalibrationRequest":
        if not self.autore_e_soggetto_coincidono:
            raise ValueError(
                "La calibrazione va usata solo con testo scritto dalla stessa persona "
                "monitorata, con il suo consenso."
            )
        return self


class ModelDiagnostics(BaseModel):
    model_config = ConfigDict(extra="allow")

    model_version: str
    turni_processati: int
    feedback_ricevuti: int
    norma_B: float
    norma_C: float
    traccia_P: float
    alpha: float
    baseline_intensita: list[float]
    baseline_saturato: bool


class CalibrationResult(ModelDiagnostics):
    frasi_processate: int
